### PR TITLE
[FIX]Error while replicating an Activities

### PR DESCRIPTION
**Pull request by jananjoy**

### CODE REVIEW

File: project_event/models/project_task.py [hunk 1]
```diff
@@ -396,20 +396,26 @@ def create_children_from_activity_create(self, children, parent_id):
 
     @api.multi
     def write(self, vals):
-        if self.is_activity():
-            return self.write_activity(vals)
-        else:
-            self.update_reservation_event(vals)
-            return super(Task, self).write(vals)
+        super(Task, self).write(vals)
```
Comment (rimbendhaou) [inline project_event/models/project_task.py]: It will write 2 times?

File: project_event/models/project_task.py [hunk 2]
```diff
@@ -396,20 +396,26 @@ def create_children_from_activity_create(self, children, parent_id):
...
+        for rec in self:
+            if rec.is_activity():
+                return rec.write_activity(vals)
+            else:
+                rec.update_reservation_event(vals)
+                return super(Task, rec).write(vals)
 
     def is_activity(self):
-        return self.activity_task_type == 'activity'
+        self.ensure_one()
```
Comment (rimbendhaou) [inline project_event/models/project_task.py]: if it is only one record why we need to use for rec in self?

**Review by rimbendhaou — CHANGES_REQUESTED** [review]

File: project_event/models/project_task.py (hunk 2 above)
Comment (jananjoy) [reply project_event/models/project_task.py]: @rimbendhaou  Changed Done

File: project_event/models/project_task.py (hunk 1 above)
Comment (jananjoy) [reply project_event/models/project_task.py]: Remove the duplicate right  with update_reservation_events because values are set to vals in value return. I call super write() method one time . I also remove  super(Task, self).write(vals) from write_activity() method

File: project_event/models/project_task.py [hunk 3]
```diff
@@ -401,46 +401,44 @@ def create_children_from_activity_create(self, children, parent_id):
 
     @api.multi
     def write(self, vals):
-        if self.is_activity():
-            return self.write_activity(vals)
-        else:
-            self.update_reservation_event(vals)
-            return super(Task, self).write(vals)
+        super(Task, self).write(vals)
```
Comment (eilst) [inline project_event/models/project_task.py]: You should return super(Task, self).write(vals)

Comment (jananjoy) [reply project_event/models/project_task.py]: what is the implication?

**Review by rimbendhaou — CHANGES_REQUESTED** [review]
It raises exception when duplicate event (This task must have a project since they are linked to timesheets)

File: project_event/models/project_task.py [hunk 4]
```diff
@@ -459,46 +459,45 @@ def create_children_from_activity_create(self, children, parent_id):
 
     @api.multi
     def write(self, vals):
-        if self.is_activity():
-            return self.write_activity(vals)
-        else:
-            self.update_reservation_event(vals)
-            return super(Task, self).write(vals)
+        update = super(Task, self).write(vals)
```
Comment (eilst) [inline project_event/models/project_task.py]: this  will cause doubleright

**Review by eilst — CHANGES_REQUESTED** [review]

File: project_event/tests/test_project_event.py [hunk 5]
```diff
@@ -46,3 +46,14 @@ def test_030_onchange_partner_id(self):
         self.project_1.partner_id = self.partner_tag2.id
         self.project_1._onchange_partner_id()
         self.assertEqual(self.project_1.client_type.name, 'Type 3')
+
+    def test_031_duplicate_activities(self):
+
```
Comment (eilst) [inline project_event/tests/test_project_event.py:51]: remove line

File: project_event/models/project_task.py [hunk 6]
```diff
@@ -459,46 +459,53 @@ def create_children_from_activity_create(self, children, parent_id):
...
-        if self.is_activity():
-            return self.write_activity(vals)
-        else:
-            self.update_reservation_event(vals)
-            return super(Task, self).write(vals)
+        update = None
+        for rec in self:
+            if rec.project_id:
+                update = super(Task, rec).write(vals)
+        for rec in self:
```
Comment (eilst) [inline project_event/models/project_task.py:466]: why 2 for loops on the same object? it can be put in one

File: project_event/models/project_task.py [hunk 7]
```diff
@@ -459,46 +459,53 @@ def create_children_from_activity_create(self, children, parent_id):
...
             vals['child_ids'] = temp
         return main_task
 
+    def write_activty_children(self, vals):
+        activity_child_vals = {}
+        if 'responsible_id' in vals:
+            activity_child_vals['responsible_id'] = vals['responsible_id']
+        if 'partner_id' in vals:
+            activity_child_vals['partner_id'] = vals['partner_id']
+
```
Comment (eilst) [inline project_event/models/project_task.py:502]: Remove space. This method was renamed write_children->write_activty_children 
 and the variable task_vals -> activity_child_vals. It's ok, but since you are making it more explicit, you should name activity_child_vals -> activity_childen_vals

File: project_event/models/project_task.py [hunk 8]
```diff
@@ -459,46 +459,53 @@ def create_children_from_activity_create(self, children, parent_id):
...
     @api.multi
     def write(self, vals):
-        if self.is_activity():
-            return self.write_activity(vals)
-        else:
-            self.update_reservation_event(vals)
-            return super(Task, self).write(vals)
+        update = None
+        for rec in self:
+            if rec.project_id:
```
Comment (eilst) [inline project_event/models/project_task.py:464]: If I understand the main changes of this PR are: 
for loop on write records:
and the checking of rec.project_id.
Right?

Could you please describe what was fixed? Normally it's ok but here I don't quite get it.  I think I more or less get it, but I'm also thinking that why check for project_id?

File: project_event/models/project_task.py [hunk 9]
```diff
@@ -459,46 +459,53 @@ def create_children_from_activity_create(self, children, parent_id):
...
-        if 'partner_id' in vals:
-            task_vals['partner_id'] = vals['partner_id']
-        for task in self.child_ids:
-            if task == self.get_main_task():
-                continue
-            if task_vals:
-                task.write(task_vals)
-
     @api.multi
-    def write_main_task(self, vals):
```
Comment (eilst) [inline project_event/models/project_task.py:490]: I see that you renamed write_main_task -> write_activity_main_task.
Ok, for me, I think this method didn't require any modification. But it's ok if you think we need to be more explicit.

File: project_event/models/project_task.py [hunk 10]
```diff
@@ -459,46 +459,53 @@ def create_children_from_activity_create(self, children, parent_id):
...
-        main_task = self.get_main_task()
+    def write_activity_main_task(self, vals):
         temp = []
         if 'task_state' in vals:
             return False
         if 'child_ids' in vals:
             temp = vals.pop('child_ids')
+        main_task = self.get_main_task()
         main_task = main_task.write(vals)
```
Comment (eilst) [inline project_event/models/project_task.py:491]: this can be improved: main_task = self.get_main_task().write(vals)

File: project_event/models/project_task.py [hunk 11]
```diff
@@ -947,6 +955,7 @@ def is_resource_booked(self):
 
     @api.multi
     def copy(self, default=None):
+        print('project task: def copy')
```
Comment (eilst) [inline project_event/models/project_task.py:958]: remove print

File: project_event/models/project_task.py [hunk 12]
```diff
@@ -512,6 +519,7 @@ def get_parent_project_id(self):
 
     @api.multi
     def update_reservation_event(self, vals):
+        self.ensure_one()
         if len(self) == 1:
```
Comment (eilst) [inline project_event/models/project_task.py:523]: len(self) == 1 is the same thing as self.ensure_one(). Right?

File: project_event/tests/test_project_event.py [hunk 13]
```diff
@@ -46,3 +46,14 @@ def test_030_onchange_partner_id(self):
         self.project_1.partner_id = self.partner_tag2.id
         self.project_1._onchange_partner_id()
         self.assertEqual(self.project_1.client_type.name, 'Type 3')
+
+    def test_031_duplicate_activities(self):
```
Comment (eilst) [inline project_event/tests/test_project_event.py:50]: Please follow agreements, the naming convention: test_040_du...

File: project_event/tests/test_project_event.py [hunk 14]
```diff
@@ -46,3 +46,14 @@ def test_030_onchange_partner_id(self):
...
+
+        new_rec = self.test_duplicate_project_1.copy()
+        self.assertEqual(new_rec.name, 'Test Duplicate Project 1 (copy)')
+        self.assertEqual(
+            len(self.test_duplicate_activity_1.child_ids), 1
+        )
+        self.assertEqual(
+            len(self.test_duplicate_activity_2.child_ids), 1
```
Comment (eilst) [inline project_event/tests/test_project_event.py:58]: Test for two, because you are not testing what you implemented. If I understand well, your changes are supposed to fix the problem: 1 event with 2 activities with multiple tasks. So test this.

File: project_event/tests/test_project_event.py [hunk 15]
```diff
@@ -46,3 +46,14 @@ def test_030_onchange_partner_id(self):
         self.project_1.partner_id = self.partner_tag2.id
         self.project_1._onchange_partner_id()
         self.assertEqual(self.project_1.client_type.name, 'Type 3')
+
+    def test_031_duplicate_activities(self):
+
+        new_rec = self.test_duplicate_project_1.copy()
+        self.assertEqual(new_rec.name, 'Test Duplicate Project 1 (copy)')
+        self.assertEqual(
+            len(self.test_duplicate_activity_1.child_ids), 1
```
Comment (eilst) [inline project_event/tests/test_project_event.py:55]: This is wrong: you should test new rec and whatever changes it makes....I think test_duplicate_activity_1. Why did you add this assert?

File: project_event/tests/test_project_event.py (hunk 14 above)
Comment (eilst) [inline project_event/tests/test_project_event.py:58]: What is the meaning of this assert

File: project_event/tests/test_project_event.py (hunk 13 above)
Comment (eilst) [inline project_event/tests/test_project_event.py:50]: You are very explicit renaming the methods:         self.write_main_task(vals) ->
        self.write_children(vals) -> self.write_activty_children(vals), thats good, you could put a more minighful and descriptive name to this test.

**Review by eilst — CHANGES_REQUESTED** [review]

File: project_event/tests/test_project_event.py (hunk 13 above)
Comment (jananjoy) [reply project_event/tests/test_project_event.py:50]: @eilst  Ok changed

File: project_event/tests/test_project_event.py (hunk 5 above)
Comment (jananjoy) [reply project_event/tests/test_project_event.py:51]: @eilst ok

File: project_event/tests/test_project_event.py (hunk 14 above)
Comment (jananjoy) [reply project_event/tests/test_project_event.py:58]: @ ????

File: project_event/models/project_task.py (hunk 11 above)
Comment (jananjoy) [reply project_event/models/project_task.py:958]: @eilst Good Catch
